[댓글 기능] 마틴 미션 제출합니다.

oeeen opened the pull request:
안녕하세요. 쏘쏘! 마틴입니다!

미션 기한이 생각했던 기간 보다 너무 짧아서 긴급하게 기능구현만 우선적으로 선정해서 PR을 보내게 되었습니다.

테스트 코드 작성에서 역시나 어려움이 있었고, 페어의 코드에 대한 이해도 없이 빠르게 테스트 코드 작성을 하다보니 여러 어려움이 있었던 것 같습니다.

현재 테스트 코드 전체를 돌리면 무수히 깨지는 상황이고 Test 클래스 별로 테스트 하면 통과하는 상황입니다.

페어인 이바와 리뷰어가 동일해서 이바의 코드를 보셔도 동일한 코드일 것입니다.

부족한 점이 많으니 거침없는 피드백 부탁드립니다! 항상 감사합니다!

seonghun127 committed:
[게시글 생성/조회/수정/삭제] 이바 미션 제출합니다. (#19)

* setting: mission complete

* docs: README.md

1. todo 목록 정리

* feat: build.gradle

1. devtools 의존성 추가

* feat: ArticleController Class

1. Article 도메인 객체 구현
2. Test 추가
3. ArticleRepository 저장기능 구현

* feat: layout

1. build.gradle 의존성 추가
2. 공통 레이아웃 추출 및 적용

* feat: ArticleDto Class

1. Create 용도의 ArticleDto 구현
2. ArticleDto를 사용하도록 Controller와 Repository 변경

* refactor: Article class

1. ArticleDto를 사용하기 위한 생성자 추가
2. id 자동 증가를 위한 static field 추가

* feat: ArticleController class

1. 조회 기능 구현

* style: css 및 templates 파일 코드 포메팅

* fix: app.css

1. 파일 오류 수정

* test: ArticleRepository class

1. 조회 및 전체조회 테스트

* fix: layout

1. thymeleaf-layout 버전에 따른 오류 수정

* fix: ArticleController class

1. createArticle 리다이렉팅 주소 수정

* feat: ArticleController class

1. 게시물 수정 기능 구현
2. ArticleDto.Update 생성
3. ArticleRepository에 update 기능 추가
4. update를 위해 Article 도메인 객체에 생성자 추가

* feat: article-edit.html

1. 생성 화면과 수정 화면에 따른 form 경로 및 메서드 조정
2. article.html에 하드코딩된 url 변경

* feat: ArticleController class

1. 삭제 기능 구현
2. ArticleRepository delete 구현

* feat: article.html

1. delete 버튼 구현

* refactor: layout

1. header 공통부분 추출

* fix: ArticleRepository class

1. 수정시 새 게시물이 추가되는 오류 수정

* style: 전체 코드 포매팅

* fix: article-edit.html

1. 수정 페이지에서 글 내용이 나오지 않는 오류 수정

* fix: article.html

1. 예시로 설정된 게시글 내용 삭제
2. 작성된 게시글 내용이 나오게끔 수정

* refactor: custom head 적용 방법 변경

* refactor: fragment 폴더 생성

* refactor: layout.html

1. main-head를 담는 th:replace 태그 추가
2. main-body를 담는 th:replace 태그 추가

* refactor: ArticleController class

1. 필드 의존 주입에서 생성자 의존 주입으로 변경

* refactor: NotFoundArticleException class 추가

* refactor: ArticleRepository class

1. find() 메서드의 이름 수정 : findById()

* style: 코드 포매팅

* refactor: Article class

1. set 메서드 제거

* fix: ArticleRepositoryTest 클래스에서 깨지는 테스트 코드 수정

* refactor: ArticleController class

1. @AllArgsContructor 를 @RequiredArgsContructor 어노테이션으로 변경
2. articleRepository 필드를 final 변경
- 불변셩을 유지하기 위해

* feat: Article class

1. 커스터마이즈한 Builder 패턴 추가

* refactor: Article class에 lombok @Builder 적용

* feat: ArticleController class

1. 게시물 추가 테스트 코드 구현

* style: ArticleRepositoryTests class 이름 변경

* fix: Article class

1. updateArticle() 메서드 추가

* refactor: 중복 head 부분 제거

* feat: ArticleController class

1. modelMapper 의존성 주입
2. modelMapper 클래스를 빈으로 등록
3. ArticleDto 클래스에 Response 중첩클래스 추가
4. ArticleController에서 article(s) 를 반환하는 모든 API의 반환 형태를 ArticleDto.Response 타입으로 변경

* style: 코드 포맷팅

* refactor: ArticleController class

1. updateArticle() 메서드 리턴 방식 변경
- 리다이렉트에서 포워딩 방식으로 변경한다
2. 그에 따른 인수 테스트 방식 수정
- 수정된 값을 확인하는 테스트로 수정한다

* style: 주석 제거

* refactor: ArticleRespository class

1. save() 메서드 리턴 타입 void로 변경
2. update() 메서드 리턴 타입 void로 변경

* refactor: ArticleController class

1. createArticle() 메서드 리턴 타입 변경
- 리다이렉트 방식에서 포워딩 방식으로 변경한다

* refactor: ArticleControllerTests class

1. 지속 테스트 가능한 코드로 리펙토링

* style: 코드 포맷팅

* fix: ArticleController class

1. createArticle(), updateArticle() api 리턴 타입 변경
- 포워드 방식에서 리다이렉트 방식으로 변경한다

2. ArticleRepository 클래스에서 save(), update() 메서드의 리턴 타입 변경
- void에서 long으로 변경한다

3. 테스트 코드 수정

* style: 코드 포맷팅

oeeen committed:
refactor: Article Service/Repository 합치기

seonghun127 committed:
[게시판 회원 기능] 이바 미션 제출합니다. (#66)

* setting: mission complete

* docs: README.md

1. todo 목록 정리

* feat: build.gradle

1. devtools 의존성 추가

* feat: ArticleController Class

1. Article 도메인 객체 구현
2. Test 추가
3. ArticleRepository 저장기능 구현

* feat: layout

1. build.gradle 의존성 추가
2. 공통 레이아웃 추출 및 적용

* feat: ArticleDto Class

1. Create 용도의 ArticleDto 구현
2. ArticleDto를 사용하도록 Controller와 Repository 변경

* refactor: Article class

1. ArticleDto를 사용하기 위한 생성자 추가
2. id 자동 증가를 위한 static field 추가

* feat: ArticleController class

1. 조회 기능 구현

* style: css 및 templates 파일 코드 포메팅

* fix: app.css

1. 파일 오류 수정

* test: ArticleRepository class

1. 조회 및 전체조회 테스트

* fix: layout

1. thymeleaf-layout 버전에 따른 오류 수정

* fix: ArticleController class

1. createArticle 리다이렉팅 주소 수정

* feat: ArticleController class

1. 게시물 수정 기능 구현
2. ArticleDto.Update 생성
3. ArticleRepository에 update 기능 추가
4. update를 위해 Article 도메인 객체에 생성자 추가

* feat: article-edit.html

1. 생성 화면과 수정 화면에 따른 form 경로 및 메서드 조정
2. article.html에 하드코딩된 url 변경

* feat: ArticleController class

1. 삭제 기능 구현
2. ArticleRepository delete 구현

* feat: article.html

1. delete 버튼 구현

* refactor: layout

1. header 공통부분 추출

* fix: ArticleRepository class

1. 수정시 새 게시물이 추가되는 오류 수정

* style: 전체 코드 포매팅

* fix: article-edit.html

1. 수정 페이지에서 글 내용이 나오지 않는 오류 수정

* fix: article.html

1. 예시로 설정된 게시글 내용 삭제
2. 작성된 게시글 내용이 나오게끔 수정

* refactor: custom head 적용 방법 변경

* refactor: fragment 폴더 생성

* refactor: layout.html

1. main-head를 담는 th:replace 태그 추가
2. main-body를 담는 th:replace 태그 추가

* refactor: ArticleController class

1. 필드 의존 주입에서 생성자 의존 주입으로 변경

* refactor: NotFoundArticleException class 추가

* refactor: ArticleRepository class

1. find() 메서드의 이름 수정 : findById()

* style: 코드 포매팅

* refactor: Article class

1. set 메서드 제거

* fix: ArticleRepositoryTest 클래스에서 깨지는 테스트 코드 수정

* refactor: ArticleController class

1. @AllArgsContructor 를 @RequiredArgsContructor 어노테이션으로 변경
2. articleRepository 필드를 final 변경
- 불변셩을 유지하기 위해

* feat: Article class

1. 커스터마이즈한 Builder 패턴 추가

* refactor: Article class에 lombok @Builder 적용

* feat: ArticleController class

1. 게시물 추가 테스트 코드 구현

* style: ArticleRepositoryTests class 이름 변경

* fix: Article class

1. updateArticle() 메서드 추가

* refactor: 중복 head 부분 제거

* feat: ArticleController class

1. modelMapper 의존성 주입
2. modelMapper 클래스를 빈으로 등록
3. ArticleDto 클래스에 Response 중첩클래스 추가
4. ArticleController에서 article(s) 를 반환하는 모든 API의 반환 형태를 ArticleDto.Response 타입으로 변경

* style: 코드 포맷팅

* refactor: ArticleController class

1. updateArticle() 메서드 리턴 방식 변경
- 리다이렉트에서 포워딩 방식으로 변경한다
2. 그에 따른 인수 테스트 방식 수정
- 수정된 값을 확인하는 테스트로 수정한다

* style: 주석 제거

* refactor: ArticleRespository class

1. save() 메서드 리턴 타입 void로 변경
2. update() 메서드 리턴 타입 void로 변경

* refactor: ArticleController class

1. createArticle() 메서드 리턴 타입 변경
- 리다이렉트 방식에서 포워딩 방식으로 변경한다

* refactor: ArticleControllerTests class

1. 지속 테스트 가능한 코드로 리펙토링

* style: 코드 포맷팅

* fix: ArticleController class

1. createArticle(), updateArticle() api 리턴 타입 변경
- 포워드 방식에서 리다이렉트 방식으로 변경한다

2. ArticleRepository 클래스에서 save(), update() 메서드의 리턴 타입 변경
- void에서 long으로 변경한다

3. 테스트 코드 수정

* style: 코드 포맷팅

* fix: Spring Data Jpa 사용

* docs: 기능 구현 목록 정리

* docs: h2 기본 설정

* docs: 기능 구현 목록 정리

* feat: 회원가입 요청 페이지 이동 및 저장

1. 회원 가입 페이지 이동 api(renderSignUpPage())를 구현한다
2. 회원 정보를 저장하는 api(createUser())를 구현한다

* feat: 회원등록 유효성 검사

* fix: 회원가입 api response 방식 변경

1. createUser() api의 리턴 방식을 리다이렉트로 변경한다
- 회원가입이 성공하면 "/login"으로 리다이렉트 된다
- 회원가입이 실패하면 "/signup"으로 리다이렉트 된다

* feat: 회원 조회 기능 구현

1. 회원을 조회하는 api(readUsers())를 구현한다
2. service 계층에서 dto와 domain 객체의 mapping 처리한다

* fix: 페이지 링크 수정

* feat: 서버에 예외처리 추가

1. email, password, name에 형식 조건을 추가한다
2. ControllerAdvice로 exception handler를 관리한다

* feat: UserControllerTest class

* docs: 기능 구현 목록 정리

* refactor: 서버에서 예외처리 리팩토링

1. 커스텀 예외를 추가한다
- 중복, 형식(이메일, 이름, 비밀번호)
2. ControllerAdvice와 ExceptionHandler를 적용한다
3. Response Status를 설정한다

* feat: 로그인, 로그아웃 기능 구현

1. Login 전용 DTO를 추가한다 (UserDTO.Login)
2. 로그인 성공 시 Session을 생성한다
3. 로그인 유무에 따라 페이지를 분기 이동한다

* fix: 페이지 요청 방식 수정

1. get 요청 시 화면, url 일관성 주기 위해 redirect방식으로 변경한다

* docs: 기능 구현 목록 정리

* feat: MyPage 이동 기능 구현

1. 마이페이지로 이동하는 api(renderMYPage())를 구현한다
2. 로그인 유무에 따라 이동 페이지를 결정한다

* fix: 로그인이 안되있을 시 로그아웃 기능 수정

1. session 정보가 존재할때만 로그아웃을 수행한다

* docs: 기능 목룩 구현 정리

* feat: 회원 정보 수정 기능 추가

1. 회원 정보 수정 api(updateUser())를 추가한다
2. My Page로 이동하는 api(renderMypage())를 수정한다
- session 정보를 확인하는 로직을 삭제한다
3. Edit My Page로 이동하는 api(renderEditMypage())를 추가한다
- renderMypage()에서 처리하던 session 관리 로직을 추가한다
4. UserDto.Reponse 클래스에 id 필드를 추가한다
5. Service 계층에 update() 메서드를 추가한다

* fix: 회원가입, 회원수정 시 예외 처리 코드 분류

1. InvalidSignUpFormException을 구현한다
2. InvalidEditFormException을 구현한다

* feat: 회원 탈퇴 기능 구현

1. 회원 정보를 삭제하는 api(deleteUser()) 추가한다
2. 테스트 코드 @AfterEach에 삭제 api를 호출하여 전체 테스트가 가능하게 한다

* feat: UserService 테스트 코드 추가

* fix: UserService 테스트 코드 수정

* refactor: ArticleService 계층 분리

* style: 코드 포맷팅

* docs: data.sql 추가

* fix: 테스트 코드 오류 수정

1. data.sql 파일 삭제

* refactor: 도메인 @Builder 어노테이션 적용 위치 수정

1. class 전체에 적용하던 @Builder 어노테이션을 생성자로 변경한다
- 생성자 파라미터에 대해서만 빌더 패턴을 적용하기 위해

* refactor: 회원 삭제 시 logout api 호출로 로직 변경

* refactor: 도메인별 패키지 분리

* refactor: 로그인 여부에 따른 이동경로 분기 설정

1. 화면 우측 상단에 이름 / Login 버튼에 대해 로그인 여부에 따라
이동하는 경로를 달리 지정한다
- 로그인 시 '/', 아닐 시 '/login'으로 이동한다

* feat: Interceptor 적용

* feat: Argument Resolver 구현

* refactor: session 정보 삭제 방법 수정

1. invalidate() 메서드를 removeAttribute() 메서드로 변경한다

* refactor: dto 클래스의 중첩 클래스 이름을 명사형으로 수정

* style: 코드 포멧팅

* feat: UserDto.Login 클래스에 validate 추가

1. 이메일과 비밀번호의 형식을 체크한다
2. InvalidLoginFormException을 추가한다

* feat: User 클래스의 필드 validation 추가

* refactor: update 수정

1. 변경할 객체를 생성하고 save()를 다시 호출했던 방식에서
변경할 객체에 update() 메세지를 보내어 수정되는
사항이 영속적으로 저장되는 방식으로 변경한다

* style: import 패키지 변경

* refactor: 로그인시 비밀번호 일치 검사 로직 변경

1. User 클래스에 checkPassword() 메서드를 추가한다
2. User 객체에게 checkPassword() 메세지를 보내 비밀번호를 확인한다

* refactor: 테스트 시 회원가입과 로그인 로직을 따로 추출

1. test code에서 setUp()에서 처리했던 회원가입, 로그인하는 로직을
UserUtilForTest 클래스로 추출한다

* feat: Attribute Converter 구현

1. User 도메인 필드의 타입에 대한 래핑 클래스를 생성한다
2. 서버에선 User 도메인 필드 타입을 모드 래핑 클래스로 적용한다
3. DB에 저장할 시 String 타입으로 저장해주는 래핑 클래스에 대한 Converter를 생성한다

* refactor: Name 클래스의 updateName 메서드 수정

1. VO인 Name 클래스의 불변을 위해 updateName() 메서드 로직을
변경될 필드 값에 대해 새로운 Name 객체로 반환하는 구조로 변경한다

* refactor: VO 클래스 필드에 final 추가

* refactor: domain 패키지 안에 vo 패키지 추가

* refactor: 회원 정보 수정 성공 테스트 리펙토링

1. 정보 수정 요청 후 리다이렉트 되는 페이지에서 수정된 정보가
잘 반영됐는지 확인하는 로직을 추가한다

* style: 코드 포맷팅

oeeen committed:
Merge remote-tracking branch 'upstream/seonghun127' into oeeen

seonghun127 committed:
feat: 게시글 회원 정보 연결

1. 다대일 매핑

seonghun127 committed:
feat: 게시글 회원 정보 연결

1. 수정, 삭제 시 회원 정보 확인

seonghun127 committed:
feat: 게시글 회원 정보 연결

1. 프론트에서 프로필을 회원 정보에 맞게 변경

seonghun127 committed:
feat: 댓글 생성 기능 추가

1. User 도메인과 Article 도메인을 연결한다

seonghun127 committed:
refactor: Article 클래스의 메서드명 변경

1. matchAuthorId() 에서 notMatchAuthorId()로 변경한다

seonghun127 committed:
feat: 댓글 조회, 수정, 삭제 기능 추가

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
refactor: @Data 어노테이션 추가

seonghun127 committed:
feat: 댓글 생성, 조회, 삭제 기능 구현

seonghun127 committed:
feat: 댓글 수정 기능 구현

seonghun127 committed:
feat: 테스트 코드 구현

seonghun127 committed:
feat: 오류 메세지 출력

ssosso reviewed (changes requested):
안녕하세요, 마틴~ 쏘쏘입니다:) (저와의 코드리뷰에서는 '님' 어미를 사용하지 않으셔도 됩니다)
이바와 코드가 거의 동일하시네요~ 동일한 피드백이 많습니다. 
페어 코딩하신 분들이 같은 리뷰어에 배정되는게 그렇게 좋지는 않네요;);;
이바와 함께 전체 테스트 코드를 실행했을 때 실패하는 부분은 확인해서 반드시 고치면 좋을 것 같습니다.
피드백 드린 부분 외에 궁금한 부분이 있으시면 말씀주세요:)~

oeeen committed:
refactor: redirect 방식 String에서 RedirectView로 통일

1. controller에서 exception 발생 시, redirect와 String으로
나누어 지던 것을 controllerAdvice에 새로운 Exception 클래스로 분리
2. index.html header 일부 추가

oeeen committed:
refactor: @AllArgsConstructor 제거

oeeen committed:
refactor: 각 Dto inner class 분리

oeeen committed:
refactor: @RequiredArgsConstructor 제거, 생성자 직접 입력

oeeen committed:
refactor: new UserSession(user) 로 변경

oeeen committed:
refactor: deleteById return type 변경

oeeen committed:
refactor: Exception 발생 시 Id Logging 추가

oeeen committed:
refactor: header 오류 수정

oeeen committed:
test: UserController, LoginController Test 재작성

oeeen committed:
refactor: production code mysql로 이전

test code 용 application.properties 추가
data.sql 추가

oeeen committed:
refactor: Article, Comment Delete cascade 추가

oeeen committed:
test: Comment Controller, domain 테스트 추가

oeeen committed:
test: User Controller, domain 테스트 추가

oeeen committed:
test: Article Controller, domain 테스트 추가

oeeen committed:
test: User service Test 추가

FindByEmail을 위한 Email Equals 추가
Save에 return type 추가

oeeen committed:
test: Article Service Test 추가

oeeen committed:
test: CommentService 테스트 추가

1. Comment Entity Equals and Hashcode 추가 (id로 구분)
2. update 메서드 return 타입 변경

oeeen committed:
style: formatting, 안쓰는 파일 삭제

oeeen commented:
안녕하세요~
테스트 코드를 전체 갈아엎고 다시 작성하느라 시간이 조금 걸렸습니다.
서비스 레이어 테스트를 위해 mockito를 공부해서 적용해보았는데, 처음 사용하는 거라 적절하게 사용하지는 못한 것 같습니다.

페어인 이바와 서로 피드백 반영 방향을 공유하면서 진행 했으나 아마 다른 방향으로 피드백을 반영한 것 같습니다!

항상 도움을 주셔서 감사합니다! :)

ssosso reviewed (approved):
마틴~ 댓글주신 부분에 대한 답변 작성했습니다:) 확인 부탁드립니다~

oeeen committed:
refactor: modelmapper 사용 변경

oeeen commented:
안녕하세요~ 쏘쏘~ 피드백 및 답변 감사합니다!!
Dto에 대한 것은 Request마다 Dto가 따로 필요하다고 생각하게 되었고, 현재 상황에서는 이를 상위클래스로 나누는 것은 오히려 과한 분리라고 생각하게 되어서 변경은 하지 않았습니다!

그리고 cqs와 void return 에 대한 쏘쏘의 의견을 말씀 해주셔서 감사합니다!
확실히 delete와 update에서 메서드에 리턴이 있으면 테스트코드 작성에 편할 것이라는 생각이 드네요!
그리고 쏘쏘가 언급해준 것처럼 메서드를 사용하는 측(현재는 Controller)에서 반환값을 이용해서 추가적인 로직을 진행 할 수 있을 것 같네요.
저는 그것을 오히려 Controller에 과도한 일을 시키는 게 아닐까 라는 생각에 모든 처리를 service에서 처리하도록 했었어요! 
저도 이바와 똑같이 Service에서 예외를 던져주면 Controller에서 그에 대한 처리를 또하면 중복이 아닐까 생각했었는데, 그게 아니라 Controller에서는 현재처럼 ControllerAdvice에서 처리 해주고, service 메서드의 return 값으로 controller에서는 다른 흐름제어나 로직을 수행할 수 있게 할 수 있게 하는 유연함을 줄 수 있겠구나 생각하게 되었습니다!

생각할만한 여지를 주셔서 감사합니다!

oeeen committed:
refactor: void return 제거
